pkg/commands/kvstoreentry/create.go

```diff
@@ -26,20 +28,233 @@ func NewCreateCommand(parent cmd.Registerer, g *global.Data, m manifest.Data) *C
+	if err != nil {
+		c.Globals.ErrLog.Add(err)
+		return err
+	}
+	return c.CallBatchEndpoint(f, out)
+}
+
+// ProcessDir concurrently reads files from the given directory structure and
+// uploads each file to the set-value-for-key endpoint where the filename is the
+// key and the file content is the value.
+//
+// NOTE: Unlike ProcessStdin/ProcessFile content doesn't need to be base64.
+func (c *CreateCommand) ProcessDir(out io.Writer) error {
+	path, err := filepath.Abs(c.dirPath)
+	if err != nil {
+		return err
+	}
+
+	files, err := os.ReadDir(path)
+	if err != nil {
+		return err
+	}
+
+	spinner, err := text.NewSpinner(out)
+	if err != nil {
+		return err
+	}
+
+	err = spinner.Start()
+	if err != nil {
+		return err
+	}
+	fileLength := len(files)
+	msg := "Processed %d of %d files"
+	spinner.Message(fmt.Sprintf(msg, 0, fileLength) + "...")
+
+	base := filepath.Base(path)
+	errors := []ProcessErr{}
+	processed := make(chan struct{}, c.dirConcurrency)
+	sem := make(chan struct{}, c.dirConcurrency)
+
+	var (
+		filesProcessed uint64
+		wg             sync.WaitGroup
+	)
+
+	go func() {
+		for range processed {
+			filesProcessed++
+			spinner.Message(fmt.Sprintf(msg, filesProcessed, fileLength) + "...")
+		}
+	}()
+
+	for _, file := range files {
+		if !file.IsDir() {
+			wg.Add(1)
+
+			go func(file fs.DirEntry) {
+				// Restrict resource allocation if concurrency limit is exceeded.
+				sem <- struct{}{}
+				defer func() {
+					<-sem
+				}()
+				defer wg.Done()
+
+				filePath := filepath.Join(c.dirPath, file.Name())
+				dir, filename := filepath.Split(filePath)
+				index := strings.Index(dir, base)
+				filename = filepath.Join(dir[index:], filename)
+
+				// G304 (CWE-22): Potential file inclusion via variable
+				// #nosec
+				fileContent, err := os.ReadFile(filePath)
+				if err != nil {
+					errors = append(errors, ProcessErr{
+						File: filePath,
+						Err:  err,
+					})
+					return
+				}
+
+				err = c.Globals.APIClient.InsertKVStoreKey(&fastly.InsertKVStoreKeyInput{
+					ID:    c.Input.ID,
+					Key:   filename,
+					Value: string(fileContent),
+				})
+				if err != nil {
+					errors = append(errors, ProcessErr{
+						File: filePath,
+						Err:  err,
+					})
+					return
+				}
+
+				processed <- struct{}{}
+			}(file)
+		}
+	}
+
+	wg.Wait()
+
+	// NOTE: We add a two second sleep to allow the final goroutine to increment.
+	// Otherwise the StopMessage below is called before filesProcessed is updated.
+	time.Sleep(time.Second * 2)
+
+	spinner.StopMessage(fmt.Sprintf(msg, filesProcessed, fileLength))
+	err = spinner.Stop()
+	if err != nil {
+		return err
+	}
+
+	outputMsg := "Inserted %d keys into KV Store"
+
+	if len(errors) == 0 {
+		text.Success(out, outputMsg, len(files))
+		return nil
+	}
+
+	text.Error(out, "Encountered the following errors")
+	text.Break(out)
+	for _, err := range errors {
+		fmt.Printf("File: %s\nError: %s\n\n", err.File, err.Err.Error())
+	}
+
+	return nil
```

Comment:
Should this return `nil`? If yes, please provide a comment.

Reply:
Fixed in [ffbb29d](https://github.com/fastly/cli/pull/927/commits/ffbb29d7c8ebe3b95c27f754cacac32d5f1b182b)